### PR TITLE
Adding cdap lcm backend metrics

### PR DESCRIPTION
**Pull request by sanjanasandeep**
JIRA : https://cdap.atlassian.net/browse/CDAP-20400

metrics to count deploy events

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java [hunk 1]
```diff
@@ -157,12 +159,15 @@ public class AppMetadataStore {
   /**
    * Static method for creating an instance of {@link AppMetadataStore}.
    */
-  public static AppMetadataStore create(StructuredTableContext context) {
-    return new AppMetadataStore(context);
+  public static AppMetadataStore create(StructuredTableContext context,
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: It is better to keep the existing constructor and add an overloaded one that takes two arguments when it is used at multiple places. The existing one then can pass null to the overloaded one. This will remove the need to change all callers to pass a null/default argument.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java [hunk 2]
```diff
@@ -558,11 +563,22 @@ public void createApplicationVersion(ApplicationId id,
       }
       fields.add(Fields.booleanField(StoreDefinition.AppMetadataStore.LATEST_FIELD, false));
       getApplicationSpecificationTable().upsert(fields);
+      // increment metric : event.deploy.new.count
+      emitMetrics(id.getNamespace(), id.getApplication(), Constants.Metrics.AppMetadataStore.EVENT_DEPLOY_NEW_COUNT);
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: Can this be moved to `ApplicationRegistrationStage` ? The only info that is missing seems to be whether it is a deploy or edit. Is it possible to change the `createApplicationVersion` method to return a status object with the necessary details? Existing callers can ignore this and the `ApplicationRegistrationStage` can use this information to emit the metrics. This will also avoid the problem of emitting metrics if there is an exception in `writeApplication `

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java (hunk 1 above)
Comment (sanjanasandeep) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: I agree, however I believe earlier had a conversation with @albertshau @chtyim that it is not a good practice to overload constructors. @albertshau Would it be fine to have one here, since it simplifies the code?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java (hunk 2 above)
Comment (sanjanasandeep) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: so in that case, do you mean a status indicating if its a new deploy/ edit deploy? So in that case we will have to inject `metricsCollectionService` in `LocalApplicationManager`. @albertshau @tivv do you have any concerns with this ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewDataSubscriberService.java [hunk 3]
```diff
@@ -110,13 +110,13 @@ protected TransactionRunner getTransactionRunner() {
   @Nullable
   @Override
   protected String loadMessageId(StructuredTableContext context) throws Exception {
-    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context, metricsCollectionService);
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewDataSubscriberService.java]: Will this also emit metrics for preview deployment?

Comment (sanjanasandeep) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewDataSubscriberService.java]: I think we should remove this

File: cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+package io.cdap.cdap.proto;
+
+/**
+ *  The deploy event type that is used for capturing metrics.
+ */
+public enum DeployEventType {
+  NEW,
+  EDIT,
+  UPGRADE
```
Comment (greeshmaswaminathan) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java]: Is the UPGRADE type set/used right now?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewApplicationManager.java [hunk 5]
```diff
@@ -97,7 +101,7 @@ public ListenableFuture<O> deploy(I input) throws Exception {
                                                    authenticationContext, artifactRepository, impersonator));
     pipeline.addLast(new CreateDatasetInstancesStage(cConf, datasetFramework, ownerAdmin, authenticationContext));
     pipeline.addLast(new ProgramGenerationStage());
-    pipeline.addLast(new ApplicationRegistrationStage(store, usageRegistry, ownerAdmin));
+    pipeline.addLast(new ApplicationRegistrationStage(store, usageRegistry, ownerAdmin, metricsCollectionService));
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewApplicationManager.java]: nit: just use the new NoOpMetricsCollectionService() inline.

**Review by greeshmaswaminathan — COMMENTED** [review]
Can a unit test be added to check that you are getting the expected metric for each case?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java (hunk 4 above)
Comment (sanjanasandeep) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java]: no it is not used, but I added it for completeness

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: The context of the earlier conversation was in a class where there were something like 8 different constructors. At that point, using the Builder pattern makes more sense. If it's just a couple constructors it's not a big deal.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java [hunk 6]
```diff
@@ -524,9 +525,9 @@ public RunRecordDetail getRun(ProgramReference programRef, String runId) {
   }
 
   @Override
-  public void addApplication(ApplicationId id, ApplicationMeta meta) throws ConflictException {
-      TransactionRunners.run(transactionRunner, context -> {
-        getAppMetadataStore(context).createApplicationVersion(id, meta);
+  public DeployEventType addApplication(ApplicationId id, ApplicationMeta meta) throws ConflictException {
+      return TransactionRunners.run(transactionRunner, context -> {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java]: indentation should be 2 spaces

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java [hunk 7]
```diff
@@ -1143,6 +1143,15 @@ public static final class AppStateStore {
       public static final String STATE_STORE_GET_LATENCY_MS = "state.store.get.latency.millis";
       public static final String STATE_STORE_SAVE_LATENCY_MS = "state.store.save.latency.millis";
     }
+
+    /**
+     * Deploy event metrics
+     */
+    public static final class AppMetadataStore {
+      public static final String EVENT_DEPLOY_NEW_COUNT = "event.app.create.count";
```
Comment (albertshau) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: nit: let's shorten this to DEPLOY_NEW_COUNT, etc.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java (hunk 4 above)
Comment (albertshau) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java]: this is a bit of a confusing enum, since upgrade is not used even though it is a metric and it is in this enum.

Instead of this enum, I think it would be better if addApplication returned the edit number. You can change updateApplicationEdit() in AppMetadataStore to return the edit number and keep returning it all the way up. Then emit the metric based on if edit number is 1 or more

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java [hunk 8]
```diff
@@ -43,24 +48,37 @@ public class ApplicationRegistrationStage extends AbstractStage<ApplicationWithP
...
     ApplicationMeta appMeta = new ApplicationMeta(applicationSpecification.getName(), input.getSpecification(),
                                                   input.getChangeDetail());
     try {
-      store.addApplication(input.getApplicationId(), appMeta);
+      DeployEventType deployEventType = store.addApplication(input.getApplicationId(), appMeta);
+      if (deployEventType.equals(DeployEventType.NEW)) {
+        // increment metric : event.deploy.new.count
+        emitMetrics(applicationId.getNamespace(), applicationId.getApplication(),
+                    Constants.Metrics.AppMetadataStore.EVENT_DEPLOY_NEW_COUNT);
+      } else {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java:80]: this means this edit count will include upgrades. Is that the intended behavior, or are they supposed to be mutually exclusive?

Comment (sanjanasandeep) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java:80]: Oh, i think the intent is to keep them separate. I am thinking I'll set the DeployEvent to UPGRADE at L633 in AppLifecycleService https://github.com/cdapio/cdap/pull/14956/files#diff-bd61e1bd8f120c6108e9b2346f1b32ab44bfc6d476617641ba2e9458166d08b4R633 and then include the check to  filter those events out. Is that good?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java (hunk 4 above)
Comment (sanjanasandeep) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java]: Should we keep the enum as per https://github.com/cdapio/cdap/pull/14956/files#r1119425214 ?

Comment (albertshau) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/DeployEventType.java:25]: that should be fine, it makes sense if it's used in a place like that

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java (hunk 8 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java:80]: it's fine to set some flag or something to say it's an upgrade, but I don't think we'll be able to set a DeployEvent in the other scenarios, since we don't know if it's a new deploy or an edit until we call store.addApplication().

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/AppDeploymentInfo.java [hunk 9]
```diff
@@ -56,6 +56,8 @@ public class AppDeploymentInfo {
   private final ChangeDetail changeDetail;
   @Nullable
   private final ApplicationSpecification deployedApplicationSpec;
+  @Nullable
+  private final Boolean isUpgrade;
```
Comment (sanjanasandeep) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/AppDeploymentInfo.java]: @albertshau Is this approach fine? Setting this flag only during upgrades - that way we can distinguish between upgrades and other deploy events

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java [hunk 10]
```diff
@@ -341,10 +341,11 @@ Map<ProgramRunId, RunRecordDetail> getRuns(ApplicationId applicationId, ProgramR
    * Creates new application if it doesn't exist. Updates existing one otherwise.
    * @param id            application id
    * @param meta          application metadata to store
+   * @return              the number of edits to the application
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java]: should mention whether adding a new application will return 0 or 1.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/AppDeploymentInfo.java (hunk 9 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/AppDeploymentInfo.java]: yes, though this doesn't need to be nullable, it can be a boolean.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java [hunk 11]
```diff
@@ -529,14 +529,14 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramReference> progr
...
    * @param appMeta the application metadata to be written
+   * @return the number of edits to the application
    * @throws IOException if failed to write app
    * @throws ConflictException if parent-version provided in the request doesn't match the latest version, do not allow
    * app to be created
    */
-  public void createApplicationVersion(ApplicationId id,
-                                       ApplicationMeta appMeta) throws IOException, ConflictException {
+  public int createApplicationVersion(ApplicationId id,
+                                                  ApplicationMeta appMeta) throws IOException, ConflictException {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: indentation

**Review by albertshau — COMMENTED** [review]
minor comments

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java [hunk 12]
```diff
@@ -70,7 +70,7 @@ public void process(ApplicationWithPrograms input) throws Exception {
     try {
       int editCount = store.addApplication(input.getApplicationId(), appMeta);
       // increment metric : event.deploy.upgrade.count
-      if (input.isUpgrade() != null && input.isUpgrade() == true) {
+      if (input.isUpgrade() == true) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java]: can just be `if (input.isUpgrade())`

**Review by albertshau — APPROVED** [review]
minor comment, otherwise lgtm
